source/slang/slang-stdlib-textures.cpp

```diff
@@ -0,0 +1,1213 @@
+#include "slang-stdlib-textures.h"
+
+#define EMIT_LINE_DIRECTIVE() sb << "#line " << (__LINE__+1) << " \"slang-stdlib-textures.h\"\n"
+
+namespace Slang
+{
+
+// Concatenate anything which can be passed to a StringBuilder
+template<typename... Ts>
+String cat(const Ts&... xs)
+{
+    return (StringBuilder{} << ... << xs);
+};
+
+//
+// Utilities
+//
+
+const auto indentWidth = 4;
+static const char spaces[] = "                    ";
+static_assert(SLANG_COUNT_OF(spaces) % indentWidth == 1);
+
+struct BraceScope
+{
+    BraceScope(const char*& i, StringBuilder& sb, const char* end = "\n")
+    :i(i), sb(sb), end(end)
+    {
+        SLANG_ASSERT(i != spaces);
+        sb << i << "{\n";
+        i -= indentWidth;
+    }
+    ~BraceScope()
+    {
+        SLANG_ASSERT(*i != '\0');
+        i += indentWidth;
+        sb << i << "}" << end;
+    }
+    const char*& i;
+    StringBuilder& sb;
+    const char* end;
+};
+
+TextureTypeInfo::TextureTypeInfo(
+    TextureTypePrefixInfo const& prefixInfo,
+    BaseTextureShapeInfo const& base,
+    bool isArray,
+    bool isMultisample,
+    BaseTextureAccessInfo const& accessInfo,
+    StringBuilder& inSB,
+    String const& inPath)
+    : prefixInfo(prefixInfo)
+    , base(base)
+    , isArray(isArray)
+    , isMultisample(isMultisample)
+    , accessInfo(accessInfo)
+    , sb(inSB)
+    , path(inPath)
+{
+    i = spaces + SLANG_COUNT_OF(spaces) - 1;
+}
+
+void TextureTypeInfo::writeFuncBody(
+    const char* funcName,
+    const String& glsl,
+    const String& cuda,
+    const String& spirv)
+{
+    BraceScope funcScope{i, sb};
+    {
+        sb << i << "__target_switch\n";
+        BraceScope switchScope{i, sb};
+        sb << i << "case cpp:\n";
+        sb << i << "case hlsl:\n";
+        sb << i << "__intrinsic_asm \"." << funcName << "\";\n";
+        if(glsl.getLength())
+        {
+            sb << i << "case glsl:\n";
+            sb << i << "__intrinsic_asm \"" << glsl << "\";\n";
+        }
+        if(cuda.getLength())
+        {
+            sb << i << "case cuda:\n";
+            sb << i << "__intrinsic_asm \"" << cuda << "\";\n";
+        }
+        if(spirv.getLength())
+        {
+            sb << i << "case spirv:\n";
+            sb << i << "return spirv_asm\n";
+            BraceScope spirvScope{i, sb, ";\n"};
+            sb << spirv << "\n";
+        }
+    }
+}
+
+void TextureTypeInfo::writeFuncDecorations(
+    const String& glsl,
+    const String& cuda)
+{
+    if(glsl.getLength())
```

Comment:
We no longer need to write target_intrinsics, right?

Reply:
Yeah, I had it this way to minimize a massive diff from the beginning while working on this, just to verify that the refactoring was correct. I also wonder if it's slightly faster to process the decorations version, but haven't confirmed.